src/EditorFeatures/CSharp/Formatting/CSharpEditorFormattingService.cs

```diff
@@ -120,11 +120,11 @@ public bool SupportsFormattingOnTypedCharacter(Document document, char ch)
             return Formatter.GetFormattedTextChanges(root, SpecializedCollections.SingletonEnumerable(formattingSpan), document.Project.Solution.Workspace, options, rules, cancellationToken);
         }
 
-        private IEnumerable<IFormattingRule> GetFormattingRules(Document document, int position)
+        private IEnumerable<IFormattingRule> GetFormattingRules(Document document, SyntaxNode root, int position)
         {
             var workspace = document.Project.Solution.Workspace;
             var formattingRuleFactory = workspace.Services.GetService<IHostDependentFormattingRuleFactoryService>();
-            return formattingRuleFactory.CreateRule(document, position).Concat(Formatter.GetDefaultFormattingRules(document));
+            return formattingRuleFactory.CreateRule(document, position).Concat(GetTypingRules(document, root, position)).Concat(Formatter.GetDefaultFormattingRules(document));
```

Comment:
What does this mean for other clients of the formatting API?  Will they now get poorer behavior if they just get ```Formatter.GetDefaultFormattingRules(document)```?

Do they need to know about this 'typing rule' as well?  If they don't have htat typing rule in their set of rules, will this cause a subpar experience for them?

Reply:
existing one already doesn't do anything in general case.

Reply:
I'm a bit confused by this.  You removed code from BaseFormattingRule.cs.  This is the code that is then shared by all four default rules in GetDefaultFormattingRules right?  So wouldn't these rules be affected by your code move?  If anyone is using these rules, is it now ok that htey don't?

For example, WrappingFormattingRule and SuppressFormattingRule both seem affected by this.  Wouldn't that affect people using this rule?

Reply:
ya. I was too. I wasn't sure why this code was there at the first place. so removed and saw that it was actually there just for this very specific case that only get hit while typing. 

it can never be hit while other normal formatting (it could theoretically hit at random time which will in turn do some random thing which is actually worse).

so I removed and put the rule in the proper place.

Reply:
> it can never be hit while other normal formatting

Why is that the case?  can you provide an explanation to help make that clear?  (both in the review and in the code).

> (it could theoretically hit at random time which will in turn do some random thing which is actually worse

It sounds then like this will lead to a behavior change.  Can you write a test demontrating the existing bad 'random thing which is actually worse' and how it gets better with your change here? 

Thanks!

> so I removed and put the rule in the proper place.

it's not clear to the reviewer why the new place is the 'proper place'.  Can you document the code effectively to explain why the new location is 'the right place'?

Reply:
> Why is that the case? can you provide an explanation to help make that clear?

since last token is only if selection is ended at the token. if this code is included in larger selection, that rule does nothing. I dont think I need to write that in the code since no point what it was doing wrong before when the code get removed. I wrote in the PR description what it is trying to do.

> It sounds then like this will lead to a behavior change. Can you write a test demontrating the existing bad 'random thing which is actually worse' and how it gets better with your change here?

tagging @jinujoseph 
I dont think we need such test.

> it's not clear to the reviewer why the new place is the 'proper place'. Can you document the code effectively to explain why the new location is 'the right place'?

because the rule matter only when smart token formatter runs on "}" and no where else.

Reply:
> I dont think we need such test.

It seems very weird to me to not include a test when you've called out that this should be an improvement that avoids some problems.  tests are our way of validating these sorts of things and for preventing regressions in the future as it can be super easy for people to not understand these subtleties and to then move things around.

I don't think resistance to adding tests is a healthy approach to code changes, especially around ones as subtle and hard to understand as formatting.   :)  Thanks!

Reply:
> because the rule matter only when smart token formatter runs on "}" and no where else.

Sorry, i don't see how that's the case.  For example, it runs when "enter" is hit.  You did change teh code for GetFormattingChangesOnReturnAsync. YOu also touched GetFormattingChangesAsync, which is called for many different "char typedChar".  Indeed, looking at this function, it looks like it is called by FormatCommandHandler on literally every typed char and then has a lot of complex logic to determine what to do.

can you please clarify how yo know that this is only run for when "}" is hit?  And, if that is hte case, can we somehow assert that so that if we violate that assumption we learn about it immediately?

Reply:
I am not against adding test. but i do against adding test just for the sake of adding test.

Reply:
for enter case, last token won't be the "}" 

also, if it breaks stuff, then that is a bug, we will fix it and add tests for it. no reasons to add zillion tests for this kind of changes. or spend zillion times on it. again tagging @jinujoe

Reply:
> I am not against adding test. but i do against adding test just for the sake of adding test.

I never asked for a test just for the sake of adding a tes.t  I asked for a test because you said that that this could lead to a (desirable) behavioral change.  Given that, i believe a test is appropriate as it would both demonstrate that that behavioral change has happened, and to help regressions.

In the context of this PR it woudl also provide a tremendous amount of *clarity* as tests can be one of the most sure-fire ways to help convey to the reviewer how the new code is actually different from the old code.  

Thanks!

Reply:
> also, if it breaks stuff, then that is a bug, we will fix it and add tests for it. 

The point is to have the test *ahead* of time, so we catch the break when te code change is made instead of catching it *afterwards*.

> no reasons to add zillion tests for this kind of changes.

I did not ask for a zillion tests.  I simply asked for a test that demonstrates *your* claim.  I think that is a rational and responsible request to make for this type of code change.  I have no idea why there is such a resistance to simply adding a test.  Nor do i understnad why there is such resistance to clarifying behavior through comments.  These are the types of feedback present on tons of PRs, and it's part and parcel of so much of the work in the Roslyn codebase.

Thanks! :)

Reply:
> for enter case, last token won't be the "}"

Yes.  But you also check for EOF.  Which i assume you can get after enter.

Regardless, this whole discussion is helping prove my case :)   This is complex and subtle and it is *not* clear just from looking how this is intended to work.  There are many callers for many user interaction codepaths, and there are many checks going on.  The fact that after going back and forth with you a *dozen* times, it's *still* not clear to my why this is correct, is a strong indication to me that we shoudl *make this code more clear*.

I think it would be reasonable to just nip this entire back-and-forth in the bud and just spend any energy that would have been put into a response here into just putting that effort into clarifying that code :)

At the end of the day, the best you'll be able to do is finally convince me that that code is correct.  And once that happens, we'll still just want to comment the code to help others understand.  After all, if i've spent dozens of minutes on this, and i still don't get it, i'm not sure how this will be for others who are even *less* familiar with formatting than i am.

Thanks!

Reply:
this passes all existing tests and I don't believe it changes any meaningful behavior of the product. without the rule, some tests failed and I believe the rule is added for those cases which by putting back to rule at smart token formatter made all those pass again, so I think i need all my due diligent on test sides. and also I manually tried typing and there is no abnormally while typing so, I also think covered common cases. sure, there can be corner cases I didn't cover. but those can be addressed if ever discovered,

Reply:
I'm having trouble reconciling:

> and I don't believe it changes any meaningful behavior of the product.

and that it could:

>  in turn do some random thing which is actually worse

If this could do something random which was actually worse, i'd really like to understand at least what that case is.  I asked for a test because i thought that would the *simplest* way to explain things.  But since you don't want to provide a test can you at least give me a description of the "random thing which is actually worse" so i can at least *understand*?  :)

Note: i'd be willing to write the test myself to help out.  But i can't write the test if i don't actually know what sort of case it is that you're referring to.

If you can help me understand this scenario better that would be **awesome**!  Thanks!

Reply:
Also, just for future reference i really want to emphasize this part:

> I asked for a test because i thought that would the simplest way to explain things.

Tests are hugely useful for the PR process because it means i don't have to guess at hte cases you're talking about.  They're literally there 'codified' as part of the PR.  So instead of trying to figure out what "random thing which is actually worse" means, i can just look at the test and go "oh... yeah... i see it now.  previously in tihs case it would do things more poorly.  but now it does things better.  great!  now i def get the sorts of issues this was helping out".  It doesn't have to be "zillions" of tests. 

It's the 'coding' equivalent of "a picture is worth a thousand words".  Effectively dozens of messages going back and forth could have likely been avoided in a *lot* less time with just a *single* test.  Def keep that in mind for future PRs and it will likely save a lot of time :D  Thanks!

Reply:
> and I said I won't add test for it since I dont believe it was doing intentional behavior but just some random behavior which was never designed to do so at the first place. if it was, at least one test should hit, but nothing got hit.

If nothing is hitting this... why don't we just remove all the code entirely?  What purpose does it serve?

Sorry, i'm even more confused by this explanation than i was at the beginning!